### ISSUE TITLE
Move openscapes hub to GitHub Auth with Teams

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/infrastructure#1222]
### Context

We currently use auth0 for the Openscapes hub, and the admins have to manage users manually. In https://github.com/2i2c-org/infrastructure/issues/1213, @erinmr asked for features to make that process easier. One way to do that is to just switch to GitHub auth and provide access via teams - so admins can manage access outside of the JupyterHub!

### Proposal

- [x] **Important** Validate that removing a user from a GitHub team actually removes their access on JupyterHub ➡️ https://github.com/2i2c-org/infrastructure/issues/1222#issuecomment-1111955667
- [ ] Implement https://github.com/jupyterhub/oauthenticator/issues/398 to satisfy ⬆️ 
- [ ] Switch Openscapes auth to use GitHub directly:
   - [ ] Figure out which GitHub team to use to authenticate
   - [ ] Allow only users from that GitHub team to authenticate into the JupyterHub

### Updates and actions

**Comment by yuvipanda** [comment]
@sgibson91 I think deleting the user from the hub admin interface also logs them out explicitly - so we can document that as the thing for people to do when they remove them from a GitHub team?

**Comment by sgibson91** [comment]
> admins have to manage users manually.... asked for features to make that process easier. One way to do that is to just switch to GitHub auth and provide access via teams - so admins can manage access outside of the JupyterHub!

I think if do that, that means that admins have to remove a user from _both_ the GitHub Team and the JupyterHub admin panel, which seems to be adding complexity rather than simplifying the process as originally requested?

**Comment by yuvipanda** [comment]
@sgibson91 ah, true. But that seems better than making everyone log in each time they want to start a server no? Or we could use this as a way to prioritize adding refresh_user functionality for GitHubOAuthenticator, and hold off until that lands.

**Comment by sgibson91** [comment]
> we could use this as a way to prioritize adding refresh_user functionality for GitHubOAuthenticator

This would be my preferred long-term solution as I think most hub admins would prefer to only need to carry out admin tasks in one place, whether that's the admin panel or GitHub Teams interface.

I already had a chat with Min in Gitter and I'm happy to put some cycles into the upstream work providing I have some guidance (I've not worked with the OAuthenticators before!)

**Comment by sgibson91** [comment]
(while this looks like I'm doing work I'm actually comatose in a hotel after eating a lot of delicious Indian food 😋)

**Comment by yuvipanda** [comment]
@sgibson91 yeah, I think doing this in OAuthenticator is the 'right' way to do it, and I think we should prioritize that now. Especially with auth / security stuff, doing it the right way is probably going to save us headaches in the long and short term.

**Comment by damianavila** [comment]
I do agree with both of you about prioritizing the upstream work. Users will be mad if we force them to log in each time.
Btw, let's use this issue to reference/link any further work.
@sgibson91, feel free to create a dedicated board if you feel this one grows enough.

**Comment by sgibson91** [comment]
@yuvipanda asked me if he could take over the upstream oauthenticator work for this issue and I agreed since I am quite swamped by admin stuff atm

**Comment by sgibson91** [comment]
I think we could just enable GitHub Auth on the openscapes hub for now and tackle refresh_user separately?

I'd be interested in @yuvipanda's estimation of time/work and then @GeorgianaElena agreed to help me with this if no one else can take the lead.

**Comment by GeorgianaElena** [comment]
Since I'm currently blocked waiting for more info about the Callysto hub I spent a bit of today documenting about the upstream side of this task, i.e. implementing `refresh_user` in `oauthenticator`. 

So, I was planning to self-assign this and come up with an initial implementation for the upstream issue or at least a plan to push this forward hopefully tomorrow.

### A short summary of the current upstream state

There is support for [`refresh_user` in JupyterHub](https://jupyterhub.readthedocs.io/en/stable/api/auth.html#jupyterhub.auth.Authenticator.refresh_user) and there are currently two opened PRs that strive to implement this for GitLab and Generic OAuthenticators. More about this in @sgibson91's comment above https://github.com/2i2c-org/infrastructure/issues/1222#issuecomment-1111955667
That is valuable work that needs reviewing.
However, the longer term solution would be to come up with a generic implementation that each authenticator could extend, as proposed in https://github.com/jupyterhub/oauthenticator/issues/398. I believe we should pursue that, while valuing the existing work proposed in the two PRs.

P.S. I've updated the top comment with the main two sub-tasks remaining of this issue.

**Comment by damianavila** [comment]
Thanks for taking the lead on this one, @GeorgianaElena!
The plan LGTM.

**Comment by yuvipanda** [comment]
Thanks a lot for taking this on, @GeorgianaElena!

**Comment by yuvipanda** [comment]
Currently blocked on https://github.com/jupyterhub/oauthenticator/pull/526 :(

**Comment by damianavila** [comment]
We briefly discussed in the sprint meeting the problem of "visibilizing" what we do upstream since @GeorgianaElena is doing a lot of work upstream of us.

**Comment by GeorgianaElena** [comment]
The next big step towards wrapping up this task this is https://github.com/jupyterhub/oauthenticator/issues/398.

But before fixing this issue, some work has been done in order to build a more solid base for it and to improve the maintainability of the `jupyterhub/oauthenticator` project. I will link some relevant PRs below for visibility and as an update (I'm very grateful for @consideRatio's work and other's on the JupyterHub team on reviewing these)

- https://github.com/jupyterhub/oauthenticator/pull/524
- https://github.com/jupyterhub/oauthenticator/pull/526
- https://github.com/jupyterhub/oauthenticator/pull/551
- https://github.com/jupyterhub/oauthenticator/pull/554
- https://github.com/jupyterhub/oauthenticator/pull/561
- https://github.com/jupyterhub/oauthenticator/pull/568
- https://github.com/jupyterhub/oauthenticator/pull/569

My plan towards https://github.com/jupyterhub/oauthenticator/issues/398 is to:
- finish up https://github.com/jupyterhub/oauthenticator/pull/569
- fixup any remaining known bugs
- release a new `oauthenticator` version containing the refactorization and hope it doesn't break anything

And after this, start implementing https://github.com/jupyterhub/oauthenticator/issues/398 because I don't think the release should get blocked by that necessarily. But let's see how things go.

**Comment by damianavila** [comment]
Pause for now waiting for feedback from the upstream team.

**Comment by yuvipanda** [comment]
Closing in favor of https://github.com/2i2c-org/infrastructure/issues/3240 which is much more tightly scoped
